File: java/src/main/java/dev/drim/relay/app/DmController.java
package dev.drim.relay.app;

import dev.drim.relay.app.DmDtos.ConversationDto;
import dev.drim.relay.app.DmDtos.DmMessageDto;
import dev.drim.relay.domain.Entities;
import dev.drim.relay.domain.Events;
import dev.drim.relay.domain.Previews;
import dev.drim.relay.id.IdFactory;
import dev.drim.relay.seams.ConversationWriter;
import dev.drim.relay.seams.DmAccess;
import dev.drim.relay.seams.NotificationJobs;
import dev.drim.relay.seams.Seams.ConversationCreateResult;
import dev.drim.relay.store.ConversationRepository;
import dev.drim.relay.store.DmMessageRepository;
import dev.drim.relay.store.UserRepository;
import dev.drim.relay.store.entity.DmMessageRow;
import dev.drim.relay.web.ApiException;
import dev.drim.relay.web.CurrentUser;
import dev.drim.relay.web.Paging;
import java.time.Instant;
import java.util.List;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class DmController {
  private final CurrentUser currentUser;
  private final UserRepository users;
  private final ConversationRepository conversations;
  private final DmMessageRepository dmMessages;
  private final DmAccess dmAccess;
  private final ConversationWriter conversationWriter;
  private final NotificationJobs jobs;
  private final IdFactory ids;

  public DmController(
      CurrentUser currentUser,
      UserRepository users,
      ConversationRepository conversations,
      DmMessageRepository dmMessages,
      DmAccess dmAccess,
      ConversationWriter conversationWriter,
      NotificationJobs jobs,
      IdFactory ids) {
    this.currentUser = currentUser;
    this.users = users;
    this.conversations = conversations;
    this.dmMessages = dmMessages;
    this.dmAccess = dmAccess;
    this.conversationWriter = conversationWriter;
    this.jobs = jobs;
    this.ids = ids;
  }

  public record CreateConversationBody(String recipientId) {}

  public record CreateMessageBody(String text) {}

  @PostMapping("/dm/conversations")
  public ResponseEntity<ConversationDto> createConversation(
      @RequestBody(required = false) CreateConversationBody body) {
    String callerId = currentUser.id();
    String recipientId = body == null ? null : body.recipientId();

    if (callerId.equals(recipientId)) {
      throw ApiException.invalid(
          "dm:recipient:self", "You cannot open a conversation with yourself.");
    }
    if (recipientId == null || users.findById(recipientId).isEmpty()) {
      throw ApiException.notFound("user:not_found", "User not found.");
    }

    String[] pair = Entities.normalizePair(callerId, recipientId);
    ConversationCreateResult result = conversationWriter.create(pair[0], pair[1]);
    HttpStatus status = result.created() ? HttpStatus.CREATED : HttpStatus.OK;
    return ResponseEntity.status(status).body(ConversationDto.of(result.conversation()));
  }

  @GetMapping("/dm/conversations")
  public Paging.Page<ConversationDto> listConversations(
      @RequestParam(required = false) String limit, @RequestParam(required = false) String before) {
    int parsedLimit = Paging.parseLimit(limit);
    if (before != null && !before.isEmpty() && !conversations.existsById(before)) {
      throw Paging.unknownBefore();
    }
    List<ConversationDto> rows =
        conversations
            .pageForUser(currentUser.id(), emptyToNull(before), fetchLimit(parsedLimit))
            .stream()
            .map(r -> ConversationDto.of(r.toDomain()))
            .toList();
    return Paging.build(rows, parsedLimit, ConversationDto::id);
  }

  @GetMapping("/dm/conversations/{id}")
  public ConversationDto getConversation(@PathVariable String id) {
    return dmAccess
        .getForParticipant(id, currentUser.id())
        .map(ConversationDto::of)
        .orElseThrow(NotFounds::conversation);
  }

  @PostMapping("/dm/conversations/{id}/messages")
  public ResponseEntity<DmMessageDto> createMessage(
      @PathVariable String id, @RequestBody(required = false) CreateMessageBody body) {
    Entities.Conversation conv =
        dmAccess.getForParticipant(id, currentUser.id()).orElseThrow(NotFounds::conversation);

    String text = body == null ? null : body.text();
    if (text == null || text.isEmpty() || text.codePointCount(0, text.length()) > 4000) {
      throw ApiException.invalid("message:text:invalid", "text must be 1–4000 chars.");
    }

    DmMessageRow row = new DmMessageRow(ids.create(), id, currentUser.id(), text, Instant.now());
    dmMessages.save(row);

    // Pinned ordering (02-api.md §2): the notification job is enqueued AFTER the message commits,
    // awaiting the broker's publisher confirmation. A publish failure here is a 500 — the message
    // stays. This avoids the worker racing an uncommitted dm_messages row into its FK.
    String recipient = conv.userLo().equals(currentUser.id()) ? conv.userHi() : conv.userLo();
    jobs.enqueue(
        new Events.NotificationJob(
            row.getId(), id, currentUser.id(), recipient, Previews.preview(text)));

    return ResponseEntity.status(HttpStatus.CREATED).body(DmMessageDto.of(row.toDomain()));
  }

  @GetMapping("/dm/conversations/{id}/messages")
  public Paging.Page<DmMessageDto> listMessages(
      @PathVariable String id,
      @RequestParam(required = false) String limit,
      @RequestParam(required = false) String before) {
    dmAccess.getForParticipant(id, currentUser.id()).orElseThrow(NotFounds::conversation);

    int parsedLimit = Paging.parseLimit(limit);
    if (before != null
        && !before.isEmpty()
        && !dmMessages.existsByConversationIdAndId(id, before)) {
      throw Paging.unknownBefore();
    }
    List<DmMessageDto> rows =
        dmMessages.page(id, emptyToNull(before), fetchLimit(parsedLimit)).stream()
            .map(r -> DmMessageDto.of(r.toDomain()))
            .toList();
    return Paging.build(rows, parsedLimit, DmMessageDto::id);
  }

  private static String emptyToNull(String s) {
    return (s == null || s.isEmpty()) ? null : s;
  }

  /** Fetch limit+1 so {@link Paging#build} can detect whether a next page exists. */
  private static org.springframework.data.domain.Limit fetchLimit(int limit) {
    return org.springframework.data.domain.Limit.of(limit + 1);
  }
}

File: typescript/src/app/dm.controller.ts
import { Body, Controller, Get, Inject, Param, Post, Query, Req, Res } from '@nestjs/common';
import type { Response } from 'express';

import { invalid, notFound, notFoundConversation } from '../apierr/apierr.js';
import type { Conversation, DmMessage } from '../domain/domain.js';
import { normalizePair, preview } from '../domain/domain.js';
import { IdFactory } from '../idgen/idgen.js';
import { build, errUnknownBefore, parseLimit } from '../paging/paging.js';
import {
  DM_ACCESS,
  NOTIFICATION_JOBS,
  CONVERSATION_WRITER,
  type DmAccess,
  type NotificationJobs,
  type ConversationWriter,
} from '../seams/seams.js';
import { Store } from '../store/store.js';
import { currentUser, type RequestWithUser } from './identity.middleware.js';

interface ConversationDto {
  id: string;
  participantIds: string[];
  createdAt: string;
}
function toConversationDto(c: Conversation): ConversationDto {
  return { id: c.id, participantIds: [c.userLo, c.userHi], createdAt: c.createdAt.toISOString() };
}

interface DmMessageDto {
  id: string;
  conversationId: string;
  senderId: string;
  text: string;
  createdAt: string;
}

@Controller()
export class DmController {
  constructor(
    private readonly store: Store,
    private readonly ids: IdFactory,
    @Inject(DM_ACCESS) private readonly dmAccess: DmAccess,
    @Inject(CONVERSATION_WRITER) private readonly writer: ConversationWriter,
    @Inject(NOTIFICATION_JOBS) private readonly jobs: NotificationJobs,
  ) {}

  @Post('dm/conversations')
  async createConversation(
    @Req() req: RequestWithUser,
    @Res() res: Response,
    @Body() body: { recipientId?: string },
  ): Promise<void> {
    const caller = currentUser(req);
    const recipientId = body.recipientId ?? '';
    if (recipientId === caller.id) {
      throw invalid('dm:recipient:self', 'You cannot open a conversation with yourself.');
    }
    const recipient = await this.store.userById(recipientId);
    if (!recipient) {
      throw notFound('user:not_found', 'User not found.');
    }
    const [lo, hi] = normalizePair(caller.id, recipient.id);
    const result = await this.writer.create(lo, hi);
    res.status(result.created ? 201 : 200).json(toConversationDto(result.conversation));
  }

  @Get('dm/conversations')
  async listConversations(@Req() req: RequestWithUser, @Query('limit') limit?: string, @Query('before') before?: string): Promise<unknown> {
    const caller = currentUser(req);
    const parsed = parseLimit(limit);
    const cursor = before ?? '';
    if (cursor !== '' && !(await this.store.conversationExists(cursor))) {
      errUnknownBefore();
    }
    const rows = await this.store.conversationsFor(caller.id, cursor, parsed + 1);
    return build(rows.map(toConversationDto), parsed, (d) => d.id);
  }

  @Get('dm/conversations/:id')
  async getConversation(@Req() req: RequestWithUser, @Param('id') id: string): Promise<ConversationDto> {
    const caller = currentUser(req);
    const conv = await this.dmAccess.getForParticipant(id, caller.id);
    if (!conv) {
      throw notFoundConversation();
    }
    return toConversationDto(conv);
  }

  @Post('dm/conversations/:id/messages')
  async createDmMessage(
    @Req() req: RequestWithUser,
    @Res() res: Response,
    @Param('id') conversationId: string,
    @Body() body: { text?: string },
  ): Promise<void> {
    const caller = currentUser(req);
    const conv = await this.dmAccess.getForParticipant(conversationId, caller.id);
    if (!conv) {
      throw notFoundConversation();
    }
    const text = body.text ?? '';
    if (text.length < 1 || [...text].length > 4000) {
      throw invalid('message:text:invalid', 'text must be 1–4000 chars.');
    }
    const msg: DmMessage = {
      id: this.ids.create(),
      conversationId,
      senderId: caller.id,
      text,
      createdAt: new Date(),
    };
    await this.store.insertDmMessage(msg);

    // Pinned ordering (02-api.md §2): the notification job is enqueued AFTER the
    // message transaction commits, awaiting the broker's publisher confirmation.
    // A publish failure here is a 500 — the message stays.
    const recipient = conv.userLo === caller.id ? conv.userHi : conv.userLo;
    await this.jobs.enqueue({
      dmMessageId: msg.id,
      conversationId,
      senderId: caller.id,
      recipientId: recipient,
      preview: preview(text),
    });
    const dto: DmMessageDto = {
      id: msg.id,
      conversationId,
      senderId: caller.id,
      text,
      createdAt: msg.createdAt.toISOString(),
    };
    res.status(201).json(dto);
  }

  @Get('dm/conversations/:id/messages')
  async getDmMessages(
    @Req() req: RequestWithUser,
    @Param('id') conversationId: string,
    @Query('limit') limit?: string,
    @Query('before') before?: string,
  ): Promise<unknown> {
    const caller = currentUser(req);
    const conv = await this.dmAccess.getForParticipant(conversationId, caller.id);
    if (!conv) {
      throw notFoundConversation();
    }
    const parsed = parseLimit(limit);
    const cursor = before ?? '';
    if (cursor !== '' && !(await this.store.dmMessageExists(conversationId, cursor))) {
      errUnknownBefore();
    }
    const rows = await this.store.dmMessages(conversationId, cursor, parsed + 1);
    const dtos: DmMessageDto[] = rows.map((m) => ({
      id: m.id,
      conversationId: m.conversationId,
      senderId: m.senderId,
      text: m.text,
      createdAt: m.createdAt.toISOString(),
    }));
    return build(dtos, parsed, (d) => d.id);
  }
}
